src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java

```diff
@@ -155,20 +159,28 @@ public ProgrammingExerciseResource(ProgrammingExerciseRepository programmingExer
             return ResponseEntity.badRequest().headers(HeaderUtil.createAlert(applicationName, "A new programmingExercise cannot already have an ID", "idexists")).body(null);
         }
 
-        if (programmingExercise.getCourse() == null) {
-            return ResponseEntity.badRequest().headers(HeaderUtil.createAlert(applicationName, "The course is not set", "courseNotSet")).body(null);
+        // Valid exercises have set either a course or an exerciseGroup
+        exerciseService.checkCourseAndExerciseGroupExclusivity(programmingExercise, ENTITY_NAME);
+
+        // Retrieve the course over the exerciseGroup or the given courseId
+        // Security mechanism: make sure that we use the values from the database and not the once which might have been altered in the client
+        Course course;
+        if (programmingExercise.hasExerciseGroup()) {
+            // Get the course over the exercise group
+            ExerciseGroup exerciseGroup = exerciseGroupService.findOneWithExam(programmingExercise.getExerciseGroup().getId());
+            course = exerciseGroup.getExam().getCourse();
+            programmingExercise.setExerciseGroup(exerciseGroup);
+        }
+        else {
+            course = courseService.findOne(programmingExercise.getCourse().getId());
+            programmingExercise.setCourse(course);
         }
 
-        // fetch course from database to make sure client didn't change groups
-        Course course = courseService.findOne(programmingExercise.getCourse().getId());
-        User user = userService.getUserWithGroupsAndAuthorities();
-        if (!authCheckService.isInstructorInCourse(course, user) && !authCheckService.isAdmin()) {
+        // Check authorization
+        if (!authCheckService.isAtLeastInstructorInCourse(course, null)) {
```

Comment:
I don't get the goal here. Why null as the user? Shouldn't fail this all the time?

Reply:
When the user is null, the check service defaults to checking the currently logged in user

Reply:
Do you think it would make sense to overload the method to have a method 
`Boolean isAtLeastInstructorInCourse(Course course){}`
I think it's more intuitive than setting user null.

However it's not a relevant thing for release but maybe you could consider this to change in the future or reason why not.

Reply:
out of scope for this PR